### ISSUE TITLE
fix: infinite loop when transforming objects with properties which return an object clone

### ISSUE THREAD

**Issue by razvanz** [typestack/class-transformer#418]
### Description

The transformer ends up into an infinite loop when transforming objects with properties which return a clone of that particular object.

I stumbled upon the problem when trying to transform a [dinero.js](https://github.com/dinerojs/dinero.js) object.

**Minimal code-snippet showcasing the problem**

```ts
import DineroFactory, { Dinero } from 'dinero.js'
import {
  Transform,
  plainToClass
} from 'class-transformer'

class ValueDto {
  @Transform((value: Dinero) => ({
    amount: value.getAmount(),
    currency: value.getCurrency()
  }))
  value: { amount: number; currency: string; }
}

const plainObj = {
  value: DineroFactory({ amount: 100, currency: 'USD' })
}

console.log(plainToClass(ValueDto, plainObj))
```

### Expected behaviour

Expect the value to be transformed and not end up in an infinite loop.

### Actual behaviour

In the example above, the transformer is traversing all properties of the Dinero object, and it gets to a setter like [`setLocale`](https://github.com/dinerojs/dinero.js/blob/develop/src/dinero.js#L157) which returns a new instance of the object configured with a different locale. The transformer then traverses the returned object which also has a `setLocale` so it runs into an infinite loop.

### Additional notes

Furthermore, I see the class transformer is traversing the property tree in order to provide the conversion, which is fine, but, as proven here, it's impossible to cover all implementation patterns in the JS world, so it would be useful to have an overwrite function, something which seems to be already reported in #96 .

**Comment by dfrancoultra** [comment]
This also kind of happens with [Decimal.js](https://github.com/MikeMcl/decimal.js/) where the default constructor is called without params and causes the constructor to throw an error.

Something like #96 is useful to give flexibility for custom implementstions.

**Comment by NoNameProvided** [comment]
This is fixed by 90feca36d2b44b6c66428ccef9402b13b58f7e2e. The below snipper works with current develop: 

```ts
import 'reflect-metadata';

import DineroFactory, { Dinero } from 'dinero.js';
import { Transform, plainToInstance } from '.';

class ValueDto {
  @Transform(({ value }) => ({
    amount: value.getAmount(),
    currency: value.getCurrency(),
  }))
  value: { amount: number; currency: string };
}

const plainObj = {
  value: DineroFactory({ amount: 100, currency: 'USD' }),
};

console.log(plainToInstance(ValueDto, plainObj));
// logs ValueDto { value: { amount: 100, currency: 'USD' } }
```
